src/zql/README.md

```diff
@@ -0,0 +1,211 @@
+
+We can jump to that set of rows rather than feeding all rows. If a limit is specified we can stop reading rows once we hit the limit.
+
+The limit functionality is implemented by making `Multiset` lazy. See `Multiset.ts` for how this is currently implemented. The limited view that is pulling values from a multiset can stop without all values being visited.
+
+Not yet implemented would be index selection. E.g., queries of the form:
+
+```ts
+Issue.where('id', '=', x);
+Issue.where('id', '>', x);
+```
+
+should just be lookups against the primary key rather than a full scan.
+
+If a view's order does not match a source's order, we will (not yet implemented here) create a new version of the source that is in the view's order. This source will be maintained in concert with the original source and used for any queries that need the given ordering.
+
+# Incremental Maintenance
+
+Incremental maintenance is simply a matter of feeding each write through the graph as it happens. The graph will produce the correct aggregate result at the end.
+
+The details of _how_ that works are specific to individual operators. For operators that only use the current row, like map & filter, it is trivial. They just emit their result. For join and reduce (not implemented yet) it is more complex.
+
+What has been implemented here lays the groundwork for join & reduce hence why it isn't as simple as a system that only needs to support map & filter.
+
+# Sources: Stateful vs Stateless
+
+Sources model tables. A source can come in stateful or stateless variants.
+
+A stateless source cannot return historical data to queries that are subscribed after the source was created.
+
+A stateful source knows it contents. When a data flow graph is attached to it, the full contents of the source is sent through the graph, effectively running the query against historical data.
+
+# Views: ValueView, TreeView
+
+There are currently two kinds of views: `ValueView` and `TreeView`.
+
+`ValueView` maintains the result of a `count` query.
+
+`TreeView` maintains `select` queries.
+
+> count and select are distinct queries. This implementation does not support returning a count with other selected columns.
+
+`TreeView` holds a comparator which uses the columns provided to `Order By` to sorts it contents. If no `Order By` is specified then the items are ordered by `id`. Any time an `Order By` is specified that does not include the `id`, the `id` is appended as the last item to order by. This is so we get a stable sort order when users sort on columns that are not unique.
+
+Views can be subscribed to if a user wishes to be notified whenever the view is updated.
+
+Aaron was pretty adamant about using native JS collections, hence the `value` property on `TreeView` returns a JS array. This is fine for cases where the view has a `limit` but for cases where you want a view of thousands+ of items I'd recommend the `PersistentTreapView` (not available here, but in Materialite).
```

Comment:
Hey, here I am. I go back and forth about this – it's possible I'm wrong. My main concern is the dx – the dx of true arrays is so nice because JS has so many language helpers and so on for them. When @arv gets in here he'll have opinions I'm sure and may overrule.

I think that in practice we do not want people querying thousands of items, they should instead page them in < 1k at a time. But I can think of use cases for querying thousands of items at once.